httpx/dispatch/connection_pool.py

```diff
@@ -124,7 +125,10 @@ def num_connections(self) -> int:
             self.active_connections.remove(connection)
             self.max_connections.release()
             raise exc
-
+        except NewConnectionRequired:
+            return await self.send(
+                request=request, verify=verify, cert=cert, timeout=timeout
+            )
```

Comment:
It seems this block will actually never run, because it is shadowed by the catch-all block for `BaseException` — we need to move it above the one for `BaseException`.